### ISSUE TITLE
bazel build error for PolymerElements

### ISSUE THREAD

**Issue by jayendra13** [tensorflow/tensorflow#452]
I am trying to build tensorflow from source, and bazel is giving some unrelated error 

```
eddie7@albus:~/lab/tensorflow$ git pull
Already up-to-date.
eddie7@albus:~/lab/tensorflow$ ~/lab/bazel/bazel-bin/src/bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
.......
ERROR: /home/eddie7/lab/tensorflow/tensorflow/tensorboard/bower/BUILD:3:1: no such package '@iron-validatable-behavior//': https://github.com/PolymerElements/iron-validatable-behavior.git: cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Loading failed; build aborted.
INFO: Elapsed time: 129.792s

```

_FYI: I have also build bazel from source_

**Comment by girving** [comment]
@danmane: Do you know what might be going wrong?

**Comment by hamidb** [comment]
I encountered the same issue after adding Android NDK to the `WORKSPACE`. 
I also built bazel from the source.
It happened after syncing my tensorflow directory with Change 109591880 that apparently fixes loading dummy shared library issue #419 .

**Comment by hamidb** [comment]
I have the same issue after adding Android NDK to the `WORKSPACE`.
It happened to me after syncing my tensorflow directory with cd53f3c3302c9312c1840389a9988a879b8b9dd5 change number 109591880 that apparently load `pthread` shared library after building it through the top level build flow. (related to issue #419)

**Comment by vrv** [comment]
@hamidb: that sounds like a red herring.

Can you try 
bazel clean

followed by

bazel build -j 1 --spawn_strategy=standalone 

(These are the instructions we've been suggested for those having trouble with bazel builds)...

**Comment by jayendra13** [comment]
@vrv 
I tried to rebuild tensorflow after, executing your instruction, still no luck

```
eddie7@albus:~/lab/tensorflow$ ~/lab/bazel/bazel-bin/src/bazel build -j 1 --spawn_strategy=standalone
.......
INFO: Found 0 targets...
INFO: Elapsed time: 2.096s, Critical Path: 0.01s
eddie7@albus:~/lab/tensorflow$ ~/lab/bazel/bazel-bin/src/bazel clean
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.
eddie7@albus:~/lab/tensorflow$ git pull
remote: Counting objects: 34, done.
remote: Total 34 (delta 28), reused 28 (delta 28), pack-reused 5
Unpacking objects: 100% (34/34), done.
From https://github.com/tensorflow/tensorflow
   f88fb6f..8242b4d  master     -> origin/master
 * [new tag]         0.6.0      -> 0.6.0
Updating f88fb6f..8242b4d
Fast-forward
 tensorflow/core/public/version.h                                | 2 +-
 tensorflow/models/image/imagenet/BUILD                          | 1 +
 tensorflow/python/BUILD                                         | 1 +
 tensorflow/python/kernel_tests/parsing_ops_test.py              | 4 ++--
 tensorflow/python/kernel_tests/sparse_serialization_ops_test.py | 2 +-
 tensorflow/python/ops/gradients.py                              | 2 +-
 tensorflow/python/ops/image_ops_test.py                         | 4 ++--
 tensorflow/tools/pip_package/setup.py                           | 2 +-
 8 files changed, 10 insertions(+), 8 deletions(-)
eddie7@albus:~/lab/tensorflow$ ./configure 
Please specify the location of python. [Default is /usr/bin/python]: 
Do you wish to build TensorFlow with GPU support? [y/N] n
No GPU support will be enabled for TensorFlow
Configuration finished
eddie7@albus:~/lab/tensorflow$ ~/lab/bazel/bazel-bin/src/bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
ERROR: /home/eddie7/lab/tensorflow/tensorflow/tensorboard/bower/BUILD:3:1: no such package '@mocha//': https://github.com/mochajs/mocha.git: cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Loading failed; build aborted.
INFO: Elapsed time: 129.614s
```
it seems issue with bazel rather than tensorflow, as far as I think.
I have never used bazel before too, so can't tell anything

**Comment by vrv** [comment]
Sorry, I meant:
$ ~/lab/bazel/bazel-bin/src/bazel build -j 1 --spawn_strategy=standalone -c opt //tensorflow/tools/pip_package:build_pip_package

That is indeed weird that it can't open the mocha.git file...

**Comment by hamidb** [comment]
@vrv I have no problem with mocha.git. It was @jayendra13's error.
I was able to fix my issue. I was trying to build the android sample. I noticed that to be able to build the android sample, you need to first build other dependencies such as protobuf, etc. 
I would ideally expect calling
 `bazel build //tensorflow/examples/android:tensorflow_demo` 
it should automatically build dependencies if already not built. Am I right?
So if you have a clean workspace, build bazel android_demo does not work.

**Comment by vrv** [comment]
Assigning to andrew to take a look.

**Comment by andrewharp** [comment]
I am unable to reproduce this issue using a fresh bazel install (from source) and clone of the TF repo. tensorflow/examples/android:tensorflow_demo still builds cleanly for me.

Note that unless you clone the TF repo with --recursive, you will not grab the protocol module into your local tree. Could this be related to your issue?

**Comment by hamidb** [comment]
I had protobuf already cloned.
I will try to reproduce it again and provide more details.

**Comment by hamidb** [comment]
@andrewharp 
I can reproduce it when using `bazel 0.1.1`. It compiles with `0.1.2`
Here are the steps to reproduce:

1- bazel clean --expunge

2- cd google/protpbuf/ & make clean

3- rm -rf ~/.cache/bazel/_bazel_user

4- compile bazel

5- ./configure tensorflow without gpu

6- bazel build --verbose_failures  //tensorflow/examples/android:tensorflow_demo

I get this:
```
ERROR: /home/hamid/.cache/bazel/_bazel_hamid/f7e3d3cb1f79b9a7f53120cc4d577bcd/external/default_android_tools/src/main/protobuf/BUILD:19:2: every rule of type gensrcjar implicitly depends upon the target '//third_party:protoc',  but this target could not be found. If this is an integration test, maybe you forgot to add a mock for your new tool?.
ERROR: Loading failed; build aborted.
```

It will be OK if I first compile protobuf and then build the demo. But it would be nicer for the demo build flow to automatically resolve it.

**Comment by andrewharp** [comment]
To confirm, you are unable to reproduce the issue when building with bazel 0.1.2?

I am unable to follow your reproduction steps for "cd google/protobuf & make clean":

~/tf_git/tensorflow$ cd google/protobuf/ & make clean
[1] 939
make: *** No rule to make target `clean'.  Stop.

Given that "//third_party:protoc" appears nowhere in the Tensorflow or google/protobuf repos, I'm guessing this sounds like a bad config is being pulled in from somewhere external, likely Bazel.

At its root, when you chase down what tensorflow/core:protos_cc actually is (this nessecitates following several layers of indirection between tensorflow/core/BUILD, tensorflow/tensorflow.bzl, tensorflow/core/platform/default/build_config/BUILD, and tensorflow/core/platform/default/build_config.bzl), you find it's a cc_proto_library from /google/protobuf/protobuf.bzl. 

Are you able to build tensorflow/core:protos_cc independently?

**Comment by hamidb** [comment]
@andrewharp 
You don't need to make clean if you are using fresh protobuf repos. It will be automatically created. And yes. With 0.1.2 it compiles OK.
I will explore more since I still can reproduce it

**Comment by andrewharp** [comment]
@ahumesky Does this look like a Bazel issue that was fixed between 0.1.1 and 0.1.2?

**Comment by andrewharp** [comment]
My best guess is that this was fixed by https://github.com/bazelbuild/bazel/commit/c74ee3784d3fd3a722f92c39805e57b4353a4509

When you clean out everything some pregenerated binary must be getting lost. I would suggest using Bazel 0.1.2 or later if that solves your problem. There are no Java protocol buffers used by the demo, so  figuring out a way to somehow force the unecessary //third_party:protoc to exist for < 0.1.2 builds after cleaning would be something of a hack, especially since the root issue has already been fixed.

**Comment by hamidb** [comment]
Yes. For me, it only happens when I use 0.1.1

**Comment by andrewharp** [comment]
@jayendra13 Are you still experiencing your original problem?

**Comment by ahumesky** [comment]
'//third_party:protoc' is referring to the protoc in bazel here:
https://github.com/bazelbuild/bazel/tree/master/third_party/protobuf
https://github.com/bazelbuild/bazel/blob/master/third_party/BUILD#L20

How were you doing this step: "4- compile bazel"? If you were doing just "compile.sh compile", then the embedded tools might not have been created
@lberki may know which commit may have fixed this issue, we've done a few different things in this area

**Comment by jayendra13** [comment]
@andrewharp , Yes I am still facing the problem, this time I also tried to install bazel from ready-made shell script instead of building it. It is still giving errors!!
```
eddie7@albus:~/lab/tensorflow$ bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
Extracting Bazel installation...
......
ERROR: /home/eddie7/lab/tensorflow/tensorflow/tensorboard/bower/BUILD:3:1: no such package '@paper-dropdown-menu//': https://github.com/PolymerElements/paper-dropdown-menu.git: cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Loading failed; build aborted.
INFO: Elapsed time: 130.648s.
```

**Comment by andrewharp** [comment]
@hamidb 
I'm trying to make sure I completely understand this thread. If I understand correctly, you originally encountered this issue:
"cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'."

What was the solution for you, if you found one, and how was this connected to the protobuf issue you also related here?

@jayendra13 
Which version of Bazel are you using?

**Comment by vrv** [comment]
@lberki, @davidzchen in case they've seen this before.

@jayendra13: try searching for "cannot open git-upload-pack"  on the web and see if you can find some resources to help.  At this point it is out of scope of the TensorFlow team.

**Comment by davidzchen** [comment]
@jayendra13 - Can you paste the output of running that bazel command with `--verbose_failures`? Are you getting any similar errors with other remote repositories? I am wondering whether this is related to bazelbuild/bazel#587.

**Comment by lberki** [comment]
`cannot open git-upload-pack` seems to happen when the host name in the `remote=` attribute of `git_repository` cannot be resolved. Try running `git clone` with the URI mentioned in the offending `git_repository` rule. My hunch is that it won't work, and then it's a network misconfiguration.

Bazel could be more explicit about it, though.

**Comment by AnishShah** [comment]
I'm getting this error while installing from source. Any help? I'm now behind any proxy system. I'm able to access the link on chrome.

```
anish@anish:~/Projects/tensorflow$ bazel build -c opt //tensorflow/tools/pip_package:build_pip_package 
ERROR: /home/anish/Projects/tensorflow/tensorflow/tensorboard/bower/BUILD:5:1: no such package '@paper_styles//': Error cloning repository: https://github.com/polymerelements/paper-styles.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-styles.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-styles.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-styles.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-styles.git: cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Loading failed; build aborted.
INFO: Elapsed time: 128.161s
```

After doing `bazel clean` and running build again.. I get different error:

```
anish@anish:~/Projects/tensorflow$ bazel build -c opt //tensorflow/tools/pip_package:build_pip_package 
ERROR: /home/anish/Projects/tensorflow/tensorflow/tensorboard/bower/BUILD:5:1: no such package '@iron_checked_element_behavior//': Error cloning repository: https://github.com/polymerelements/iron-checked-element-behavior.git: cannot open git-upload-pack caused by https://github.com/polymerelements/iron-checked-element-behavior.git: cannot open git-upload-pack caused by https://github.com/polymerelements/iron-checked-element-behavior.git: cannot open git-upload-pack caused by https://github.com/polymerelements/iron-checked-element-behavior.git: cannot open git-upload-pack caused by https://github.com/polymerelements/iron-checked-element-behavior.git: cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Loading failed; build aborted.
INFO: Elapsed time: 129.496s
```

Please help.

**Comment by nitnelave** [comment]
Hi

I'm trying to build a docker image on DockerHub, and I get the same error. For the full log:

https://hub.docker.com/r/nitnelave/tensorflow/builds/bbhn9ycakwkbmzyqupkzjbh/

The dependencies are installed from https://hub.docker.com/r/nitnelave/tensorflow-dependencies/builds/bujqarnnsp7hmbetck5a6ug/ , so Bazel 0.2.1

It works on my machine, but only randomly on DockerHub (and then it times out, but that's another issue).

Sadly, I have no control on the DockerHub settings, so I have no more information than you.

**Comment by davidzchen** [comment]
+cc @kchodorow

@AnishShah IIRC the problem with using the git_* rules behind a proxy has been fixed in the most recent versions of Bazel. Which version are you using?

@nitnelave It seems that your build was able to successfully clone a bunch of the other repositories and this particular failure seems to be caused by spurious errors when cloning a repository. AFAIK, the git_repository rules currently do not support retrying failures.

**Comment by nitnelave** [comment]
@davidzchen It could be useful to have a retry, or to find what is causing the problem in the cloning process... I can't reproduce it on my machine, but it's annoying to have the build system fail every other build because of this.

**Comment by davidzchen** [comment]
@nitnelave Agreed, might be good to see if we jgit supports retries. Can you open a bug on bazelbuild/bazel for the Git retry issue?

@AnishShah If you are using an old version of Bazel, can you try using the latest version and see if that works for you?

**Comment by lightingghost** [comment]
same problem here. Using the newest tf and bazel. The error message is as following:

`no such package '@paper_dialog//': Error cloning repository: https://github.com/polymerelements/paper-dialog.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-dialog.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-dialog.git: cannot open git-upload-pack and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Loading failed; build aborted.
`

**Comment by davidzchen** [comment]
@lightingghost Can you paste the full output from running your `bazel` command with `--verbose_failures`?

**Comment by lightingghost** [comment]
@davidzchen Thanks for helping.

I am using bazel 0.2.2b at that time, when I changed to 0.2.2, it works fine.

**Comment by girving** [comment]
@davidzchen: What's the status of this?

**Comment by newcoder** [comment]
experiencing same issue with bazel 0.3.0:

ERROR: /home/ray/tensorflow/tensorflow/tensorboard/bower/BUILD:5:1: no such package '@paper_radio_group//': Error cloning repository: https://github.com/polymerelements/paper-radio-group.git: cannot open git-upload-pack caused by https://github.com/polymerelements/paper-radio-group.git: cannot open git-upload-pack caused by Connection reset and referenced by '//tensorflow/tensorboard/bower:bower'.
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted.
INFO: Elapsed time: 19.680s

**Comment by tatatodd** [comment]
@davidzchen Friendly ping, do you have any update on this?

**Comment by davidzchen** [comment]
Sorry for the wait. This fell off my radar, and I just returned from vacation.

bazelbuild/bazel#1408 is the tracking issue for Git rules written in Skylark. @nelhage has already committed an initial implementation of the [Skylark Git rules](https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/git_repositories.bzl). There is still some work to do before we can fully migrate over, but in the meantime, you can give them a try by `load()`ing `@bazel_tools//tools/build_defs/repo:git_repositories.bzl`. These rules should be drop-in replacements for the Java ones and use the Git binaries installed on your system rather than jgit.

Note that it seems the Git binaries [also do not support retries](http://stackoverflow.com/questions/35014012/git-retry-if-http-request-failed). If you are still having the same issue with the Skylark Git rules, then we should look into making use of [`git-retry`](https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-retry.html), which is included with Chromium's `depot_tools` and not the standard Git distribution.

**Comment by aselle** [comment]
Closing due to lack of recent activity. Please open a new issue referencing this one if it is still an issue. Thanks!
